controllers/rsyncnodedatamovement_controller.go

```diff
@@ -106,59 +144,98 @@ func (r *RsyncNodeDataMovementReconciler) Reconcile(ctx context.Context, req ctr
 
-	out, err := cmd.Output()
+		var cmd *exec.Cmd
+		if rsyncNode.Spec.Simulate != "" {
+			log.Info("Simulating Data Movement", "command", rsyncNode.Spec.Simulate)
+			cmd = exec.CommandContext(ctxCancel, "sh", "-c", rsyncNode.Spec.Simulate)
+		} else {
+			cmd = exec.CommandContext(ctxCancel, "rsync", arguments...)
+		}
 
-	if err != nil {
-		if exitErr, ok := err.(*exec.ExitError); ok {
-			log.Info("Rsync failure", "error", string(exitErr.Stderr))
+		// Set the appropriate permissions
+		if rsyncNode.Spec.UserId != uint32(os.Getuid()) || rsyncNode.Spec.GroupId != uint32(os.Getgid()) {
+			cmd.SysProcAttr = &syscall.SysProcAttr{
+				Credential: &syscall.Credential{
+					Uid: rsyncNode.Spec.UserId,
+					Gid: rsyncNode.Spec.GroupId,
+				},
+			}
+		}
+		log.Info("Rsync credential", "uid", rsyncNode.Spec.UserId, "gid", rsyncNode.Spec.GroupId)
+
+		// Run the command and wait for completion
+		log.Info("Executing rsync command", "source", source, "destination", destination)
+		out, cmdErr := cmd.CombinedOutput()
+		cmd.Wait()
+
+		// Log output based on result
+		if errors.Is(ctxCancel.Err(), context.Canceled) {
+			log.Info("Rsync cancelled", "error", cmdErr, "output", string(out))
+		} else if cmdErr != nil {
+			log.Info("Rsync failure", "error", cmdErr, "output", string(out))
 		} else {
-			log.Info("Rsync failure", "error", err, "output", string(out))
+			log.Info("rsync completed", "output", string(out))
 		}
-	} else {
-		log.Info("rsync completed", "output", string(out))
-	}
 
-	// Record the completion status.
-	rsyncNode.Status.EndTime = metav1.NowMicro()
-	rsyncNode.Status.State = nnfv1alpha1.DataMovementConditionTypeFinished
+		// Record the completion status and retry on conflicts
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			rsyncNode = &dmv1alpha1.RsyncNodeDataMovement{}
+			if err := r.Get(ctx, req.NamespacedName, rsyncNode); err != nil {
+				return err
+			}
 
-	if err != nil {
-		rsyncNode.Status.Status = nnfv1alpha1.DataMovementConditionReasonFailed
-		rsyncNode.Status.Message = err.Error()
-	} else {
-		rsyncNode.Status.Status = nnfv1alpha1.DataMovementConditionReasonSuccess
-	}
+			rsyncNode.Status.EndTime = metav1.NowMicro()
+			rsyncNode.Status.State = nnfv1alpha1.DataMovementConditionTypeFinished
 
-	if err := r.Status().Update(ctx, rsyncNode); err != nil {
-		log.Error(err, "failed to update rsync status with completion")
-		r.recordCompletion(*rsyncNode)
-		return ctrl.Result{}, err
+			if errors.Is(ctxCancel.Err(), context.Canceled) {
+				rsyncNode.Status.Status = nnfv1alpha1.DataMovementConditionReasonCancelled
+			} else if cmdErr != nil {
+				rsyncNode.Status.Status = nnfv1alpha1.DataMovementConditionReasonFailed
+				rsyncNode.Status.Message = cmdErr.Error()
+			} else {
+				rsyncNode.Status.Status = nnfv1alpha1.DataMovementConditionReasonSuccess
+			}
+
+			return r.Status().Update(ctx, rsyncNode)
+		})
+		if err != nil {
```

Comment:
This poses an interesting problem if we continuously get resource conflicts the resource would be abandoned. Not sure how this would work

Reply:
Do we need to return an err to the reconciler in that case? Presumably we can't sit there and retry forever, but do we need to increase the number of retries?

Reply:
This is likely so improbable I think we'd be fine with the DefaultRetry

But, if it does error for resource conflict, that was the point of the completion tracking. Nothing's really changed from the original use - you've just made it more likely (99.999999% ?) that the status would update.

I think this is an acceptable risk and the completion stuff can be removed.

@roehrich-hpe Hey short timer, what do you think?

Reply:
Ah, so I should still be recording completion even though we have error here? I ended up moving that down for only when we've successfully updated the resource.

Reply:
If RetryOnConflict() returns an error, then log it and return from the Reconcile() with a requeue, maybe with a delay rather than an immediate requeue.
I would be surprised if we ever see this, but it's simple enough to return with the requeue.

Reply:
So in this case, the RetryOnConflict() is being called in the goroutine that performs the data movement and status update. Unless I'm missing something, we don't have a way to requeue here.

Reply:
> Ah, so I should still be recording completion even though we have error here? I ended up moving that down for only when we've successfully updated the resource.

Yes, the completions map is for tracking completed operations that couldn't update the status due to resource conflicts. That hasn't changed, you've just made it extremely unlikely we would ever get a resource conflict because of the retry logic.

> So in this case, the RetryOnConflict() is being called in the goroutine that performs the data movement and status update. Unless I'm missing something, we don't have a way to requeue here.

Yeah, we've already returned the Reconcile() result.

Reply:
Ok. I believe I have a fix for this (see latest commit) to ensure that when this happens, the completions map should take care of it the next time the reconciler is ran. But the question is - what makes that reconciler run again for that request should it fail?

Can we trigger an requeue in another way that doesn't require returning the Reconcile() result?

Reply:
I don't think we can trigger a requeue easily - although the controller.Queue is accessible - so maybe after the builder runs we could hand over the controller to the reconciler then do a r.Controller.Queue.Requeue?

Reply:
Okay, I'm caught up on this.  I think it's unlikely that we won't be able to do the update.  Maybe only if someone rips the resource out from under us, and you'll catch that with the Get().

I don't think we need to get cozy with the library to figure out how to insert a requeue--let's do that only if we know it's a problem we have to solve.  At the point where we know we had to orphan it, you could bump a metric specific to that case, so we have proof that this actually happens.